### PR TITLE
[GLUTEN-7749][VL] Trim ISOControl characters in string for casting to integral type

### PR DESCRIPTION
**Pull request by wForget**
## What changes were proposed in this pull request?

Fixes: #7749, to align with Spark's change introduced by https://github.com/apache/spark/pull/41535.

## How was this patch tested?

added unit test

### CODE REVIEW

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala [hunk 1]
```diff
@@ -723,30 +723,28 @@ class VeloxSparkPlanExecApi extends SparkPlanExecApi {
...
-            c.withNewChildren(Seq(trimNode)).asInstanceOf[Cast]
-          case _ =>
-            c
+    if (GlutenConfig.getConf.castFromVarcharAddTrimNode && c.child.dataType == StringType) {
+      val trimStr = c.dataType match {
+        case BinaryType | _: ArrayType | _: MapType | _: StructType | _: UserDefinedType[_] =>
+          None
+        case FloatType | DoubleType | _: DecimalType =>
+          Some(trimSpaceStr)
+        case ByteType | ShortType | IntegerType | LongType =>
```
Comment (jackylee-ch) [inline backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala:735]: We need use `trimWhitespaceStr + trimSpaceSepStr + trimLineSepStr + trimParaSepStr + isoControlControlStr` for all other types, as they all use `UTF8String.isWhitespaceOrISOControl` to skip the invalid characters.

Comment (wForget) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala]: Makes sense, changed, thank you.

**Comment by wForget** [comment]
Run Gluten Clickhouse CI on x86

**Review by jackylee-ch — APPROVED** [review]
Basically look good to me. @PHILO-HE any more comments?

**Review by PHILO-HE — COMMENTED** [review]
@wForget, thanks for your fix! Could you give me some details about where these control chars are skipped when casting to integral types in Spark? I only found white space characters are checked and skipped.

**Comment by wForget** [comment]
> @wForget, thanks for your fix! Could you give me some details about where these control chars are skipped when casting to integral types in Spark? I only found white space characters are checked and skipped.

Introduced from [#41535](https://github.com/apache/spark/pull/41535), the relevant calls are as follows:

https://github.com/apache/spark/blob/36410f073cb978fc504f85fb25b4942dac10db3f/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L785

https://github.com/apache/spark/blob/36410f073cb978fc504f85fb25b4942dac10db3f/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L1617

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala [hunk 2]
```diff
@@ -723,30 +723,29 @@ class VeloxSparkPlanExecApi extends SparkPlanExecApi {
     val trimParaSepStr = "\u2029"
     // Needs to be trimmed for casting to float/double/decimal
     val trimSpaceStr = ('\u0000' to '\u0020').toList.mkString
+    // ISOControl characters, refer java.lang.Character.isISOControl(int)
+    val isoControlControlStr = (('\u0000' to '\u001F') ++ ('\u007F' to '\u009F')).toList.mkString
```
[the remark this answers is not in the document]
Comment (wForget) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala]: thanks, fixed

**Comment by PHILO-HE** [comment]
@wForget, I see. What I checked is Spark-3.3.1, which doesn't cover that change.

Let's note this is not applicable to all supported Spark versions. But I think it may be acceptable to end users.

**Review by PHILO-HE — APPROVED** [review]

**Pull request merged**